src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+    >>= fun i ->
+    let rec aux acc = function
+      | ((x,g)::xs) -> if i < acc+x then g else aux (acc+x) xs
+      | _ -> failwith "QCheck2.frequency crashed for an unknown reason"
+    in
+    aux 0 l
+
+  let frequencyl (l : (int * 'a) list) : 'a t =
+    List.map (fun (weight, value) -> (weight, pure value)) l
+    |> frequency
+
+  let frequencya a = frequencyl (Array.to_list a)
+
+  let char_range ?(origin : char option) (a : char) (b : char) : char t =
+    (int_range ~origin:(Char.code (Option.value ~default:a origin)) (Char.code a) (Char.code b)) >|= Char.chr
+
+  let random_binary_string (length : int) (st : Random.State.t) : string =
+    (* 0b011101... *)
+    let s = Bytes.create (length + 2) in
+    Bytes.set s 0 '0';
+    Bytes.set s 1 'b';
+    for i = 0 to length - 1 do
+      Bytes.set s (i+2) (if RS.bool st then '0' else '1')
+    done;
+    Bytes.unsafe_to_string s
+
+  let int32 : int32 t = fun st ->
+    let x = random_binary_string 32 st |> Int32.of_string in
+    Tree.make_primitive (int32_towards 0l) x
+
+  (** TODO Not actually unsigned, fix in https://github.com/c-cube/qcheck/issues/105 *)
+  let ui32 : int32 t = int32
+
+  let int64 : int64 t = fun st ->
+    let x = random_binary_string 64 st |> Int64.of_string in
+    Tree.make_primitive (int64_towards 0L) x
+
+  (** TODO Not actually unsigned, fix in https://github.com/c-cube/qcheck/issues/105 *)
+  let ui64 : int64 t = int64
+
+  let list_size (size : int t) (gen : 'a t) : 'a list t =
+    size >>= fun size ->
+    let rec loop n =
+      if n <= 0
+      then pure []
+      else liftA2 (List.cons) gen (loop (n - 1))
+    in
+    loop size
+
+  let list (gen : 'a t) : 'a list t = list_size nat gen
+
+  let list_repeat (n : int) (gen : 'a t) : 'a list t = list_size (pure n) gen
+
+  let array_size (size : int t) (gen : 'a t) : 'a array t =
+    (list_size size gen) >|= Array.of_list
+
+  let array (gen : 'a t) : 'a array t = list gen >|= Array.of_list
+
+  let array_repeat (n : int) (gen : 'a t) : 'a array t = list_repeat n gen >|= Array.of_list
+
+  let rec flatten_l (l : 'a t list) : 'a list t =
+    match l with
+    | [] -> pure []
+    | gen :: gens -> liftA2 List.cons gen (flatten_l gens)
+
+  let flatten_a (a : 'a t array) : 'a array t =
+    Array.to_list a |> flatten_l >|= Array.of_list
+
+  let flatten_opt (o : 'a t option) : 'a option t =
+    match o with
+    | None -> pure None
+    | Some gen -> opt gen
+
+  let flatten_res (res : ('a t, 'e) result) : ('a, 'e) result t =
+    match res with
+    | Ok gen -> gen >|= Result.ok
+    | Error e -> pure (Error e)
+
+  let shuffle_a (a : 'a array) : 'a array t = fun st ->
+    let a = Array.copy a in
+    for i = Array.length a-1 downto 1 do
+      let j = Random.State.int st (i+1) in
+      let tmp = a.(i) in
+      a.(i) <- a.(j);
+      a.(j) <- tmp;
+    done;
+    Tree.pure a
+
+  let shuffle_l (l : 'a list) : 'a list t = fun st ->
+    let a = Array.of_list l in
+    let _ = shuffle_a a st in
```

Comment:
Implementation was actually broken since now `shuffle_a` no longer mutates the array, but returns a copy: I fixed it in 0586f4303ec5ee39b00206a61be5d9aec6639693

Reply:
and that's why `a; b` is better than `let _ = a in b` ! :grin: 
(at least with the let it's better to use `let () = a in b` if one insists on it)